# Data Cleaning

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data= pd.read_csv('IT Salary Survey EU 2018.csv')

In [ ]:
data.head()


Timestamp   Age  ... Company size   Company type
0  14/12/2018 12:41:33  43.0  ...     100-1000        Product
1  14/12/2018 12:42:09  33.0  ...       50-100        Product
2  14/12/2018 12:47:36  32.0  ...        1000+        Product
3  14/12/2018 12:50:15  25.0  ...        1000+        Product
4  14/12/2018 12:50:31  39.0  ...     100-1000  Ecom retailer

[5 rows x 14 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 14 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Timestamp                           765 non-null    object 
 1   Age                                 672 non-null    float64
 2   Gender                              751 non-null    object 
 3   City                                736 non-null    object 
 4   Position                            737 non-null    object 
 5   Years of experience                 732 non-null    float64
 6   Your level                          743 non-null    object 
 7   Current Salary                      750 non-null    float64
 8   Salary one year ago                 596 non-null    float64
 9   Salary two years ago                463 non-null    float64
 10  Are you getting any Stock Options?  742 non-null    object 
 11  Main language at work               750 non-n

### Remove NaN values

In [ ]:
data=data.fillna('')


### 1. Leave only necessary columns

In [ ]:
data2= data[['Age','Gender','City','Position','Years of experience','Your level','Current Salary','Salary one year ago','Salary two years ago']]

In [ ]:
data2.head()

Age Gender     City  ... Current Salary Salary one year ago Salary two years ago
0  43      M  München  ...          77000               76200                68000
1  33      F  München  ...          65000               55000                55000
2  32      M  München  ...          88000               73000                54000
3  25      M  München  ...          78000               55000                45000
4  39      M  München  ...          69000               60000                52000

[5 rows x 9 columns]

### 2. Filter by columns

In [ ]:
data2.info

<bound method DataFrame.info of     Age Gender     City  ... Current Salary Salary one year ago Salary two years ago
0    43      M  München  ...          77000               76200                68000
1    33      F  München  ...          65000               55000                55000
2    32      M  München  ...          88000               73000                54000
3    25      M  München  ...          78000               55000                45000
4    39      M  München  ...          69000               60000                52000
..   ..    ...      ...  ...            ...                 ...                  ...
760  40      M     Köln  ...          44000               40000                40000
761          M     Köln  ...          45000               40000                40000
762          M     Köln  ...          45000               40000                40000
763                      ...                                                        
764  31      F  München  ...     

After filtering the rows with empty columns we have 368 rows left. The original data set contained 765 rows

### Position Column

In [ ]:
data2['Position'].unique()

array(['QA Ingenieur', 'Senior PHP Magento developer ',
       'Software Engineer', 'Senior Frontend Developer', 'UX Designer',
       'Project Manager', 'Senior Android Developer',
       'Data science team lead', 'Test Automation Engineer',
       'Frontend Developer', 'Senior Full-stack .Net Developer',
       'QA Lead', 'Front-end engineer', 'Software Developer',
       'QA Test Engineer', 'QA Manager', 'CTO', 'Cloud Consulting', 'QA',
       'Frontend Engineer', 'Head of IT', 'Senior Software Engineer',
       'Operations manager', 'Product Manager',
       'Software engineer ( React/React-native )', 'iOS Developer ',
       'Senior DevOps', 'C# Developer', 'ERP/BI Specialist',
       'Full stack developer', 'Systems Administrator',
       'Senior Software Developer', 'C++ developer', 'PM',
       'Android Developer', 'Software Architect',
       'Embedded Software Engineer', 'iOS developer', 'Data Engineer',
       'iOS Developer', 'PHP Developer', 'Ruby developer',
       'Embed

Sort the IT positions of the workers into 5 categories 

1. Backend Developers 
2. Frontend Developers
3. Full stack Developers
4. Data Scientists & Engineers
5. UX designers 

In [ ]:
fudev=['FullStack_Developer','Full Stack','fullstack','Full stack','Fullstack','Full-stack','fullsteck']
bdev= ['Devops Engineer','Senior software engineer ','System Engineer','Senior software engineer','Computer Vision','Mobile developer','SharePoint Developer','Android dev','C++ sr. s/w dev','Java','Android','Oracle Pl/SQL Developer','Senior .Net developer','Go Developer','Senior PHP developer','Hardware','SharePoint','ios','C++','Scala','IOS','Backend',' .Net Developer','Senior/Lead Developer','Scala Developer','Mobile App Developer','PHP Developer','Android Developer','C++ developer','BackEnd_Developer','Software','Java','QA','SW','Automation QA','Embedded engineer','Senior Software Engineer','DevOps','Senior DevOps','Test Automation Engineer','Senior PHP Magento developer ','E.g. C# Developer','Low level software developer','QA Ingenieur','Ruby','ruby','Java Developer ','C# Developer ','DevOps Architect ','Quality engineer','Senior Automation Test Engineer','Product Security Engineer','Lead Java Developer','Solution Engineer','Software Engineer','Erlang Developer','Oracle Developer','Lead developer','Softwaredeveloper','Senior Automation Test Engineer','System Architect','.NET','iOS','C++ Software','Python','python','Pyhon','Senior Automation Test Engineer','backend','Golang','Computer Vision Intern','Php','C#','Senior Java Developer','Java Developer','IT Engineer','.net developer','.net c# ','#c','Backend Developer']
fdev=['Senior JS Enigneer','Front-end','Sr. JS Developer','web','FrontEnd_Developer','Front End Developer ','Front End developer','Front End Engineer','Front-end developer ','FrontEnd Developer','Frontent Developer','FrontEnd dev','Front-end developer','webdev','Senior Android Developer','Frontend','Web Developer','frontend','Javascript Developer']
datas=['Senior SQL Server DBA/Developer','Business Analyst','Data Scientist','Data','data','Data science','data scientist','Data Science','Data analyst','Data Developer','Data Processing']
ux=['UX_Designer','UX_Designer','Designer','UX','ux','Ux','UI Designer','UX/UI designer','Graphic Designer','Designer','UI Designer','PCB Designer']

In [ ]:
data2['Position'].unique()

array(['QA Ingenieur', 'Senior PHP Magento developer ',
       'Software Engineer', 'Senior Frontend Developer', 'UX Designer',
       'Project Manager', 'Senior Android Developer',
       'Data science team lead', 'Test Automation Engineer',
       'Frontend Developer', 'Senior Full-stack .Net Developer',
       'QA Lead', 'Front-end engineer', 'Software Developer',
       'QA Test Engineer', 'QA Manager', 'CTO', 'Cloud Consulting', 'QA',
       'Frontend Engineer', 'Head of IT', 'Senior Software Engineer',
       'Operations manager', 'Product Manager',
       'Software engineer ( React/React-native )', 'iOS Developer ',
       'Senior DevOps', 'C# Developer', 'ERP/BI Specialist',
       'Full stack developer', 'Systems Administrator',
       'Senior Software Developer', 'C++ developer', 'PM',
       'Android Developer', 'Software Architect',
       'Embedded Software Engineer', 'iOS developer', 'Data Engineer',
       'iOS Developer', 'PHP Developer', 'Ruby developer',
       'Embed

In [ ]:
positions=['UX_Designer','FullStack_Developer','BackEnd_Developer','FrontEnd_Developer','Data Scientist']

In [ ]:
for i in range(len(data2)):
  print(i)
  for x in range(len(ux)):
    if ux[x] in data2['Position'][i]:
      data2['Position'][i]="UX_Designer"
      print(data2['Position'][i])

  for y in range(len(fudev)):
    if fudev[y] in data2['Position'][i]:
      data2['Position'][i]="FullStack_Developer"
      print(data2['Position'][i])

  for t in range(len(bdev)):
    if bdev[t] in data2['Position'][i]:
      data2['Position'][i]="BackEnd_Developer"
      print(data2['Position'][i])
  
  for p in range(len(fdev)):
    if fdev[p] in data2['Position'][i]:
      data2['Position'][i]="FrontEnd_Developer"
      print(data2['Position'][i])
  
  for w in range(len(datas)):
    if datas[w] in data2['Position'][i]:
      data2['Position'][i]="Data Scientist"
      print(data2['Position'][i])




0
BackEnd_Developer
1
BackEnd_Developer
2
BackEnd_Developer
3
FrontEnd_Developer
4
UX_Designer
UX_Designer
UX_Designer
5
6
BackEnd_Developer
BackEnd_Developer
7
Data Scientist
8
BackEnd_Developer
9
FrontEnd_Developer
10
BackEnd_Developer
11
FullStack_Developer
12
BackEnd_Developer
13
FrontEnd_Developer
FrontEnd_Developer
14
BackEnd_Developer
15
BackEnd_Developer
16
BackEnd_Developer
17
BackEnd_Developer
18
FrontEnd_Developer
19
20
21
BackEnd_Developer
22
FrontEnd_Developer
23
24
BackEnd_Developer
25
26
FrontEnd_Developer
27
28
BackEnd_Developer
29
BackEnd_Developer
30
BackEnd_Developer
31
BackEnd_Developer
32
BackEnd_Developer
33
34
FullStack_Developer
35
36
BackEnd_Developer
37
BackEnd_Developer
BackEnd_Developer
38
39
40
BackEnd_Developer
BackEnd_Developer
41
BackEnd_Developer
42
FrontEnd_Developer
43
BackEnd_Developer
44
BackEnd_Developer
45
Data Scientist
46
BackEnd_Developer
47
BackEnd_Developer
48
BackEnd_Developer
BackEnd_Developer
49
BackEnd_Developer
50
FrontEnd_Developer
51
B

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


FrontEnd_Developer
130
BackEnd_Developer
BackEnd_Developer
131
BackEnd_Developer
132
BackEnd_Developer
BackEnd_Developer
133
134
BackEnd_Developer
135
136
137
FullStack_Developer
138
139
Data Scientist
Data Scientist
Data Scientist
140
FullStack_Developer
141
FrontEnd_Developer
142
BackEnd_Developer
BackEnd_Developer
143
BackEnd_Developer
144
BackEnd_Developer
BackEnd_Developer
145
BackEnd_Developer
146
BackEnd_Developer
147
BackEnd_Developer
BackEnd_Developer
148
UX_Designer
UX_Designer
149
BackEnd_Developer
BackEnd_Developer
150
151
152
153
BackEnd_Developer
154
155
BackEnd_Developer
156
157
Data Scientist
Data Scientist
158
159
BackEnd_Developer
BackEnd_Developer
160
FullStack_Developer
161
BackEnd_Developer
162
BackEnd_Developer
BackEnd_Developer
163
BackEnd_Developer
164
Data Scientist
165
BackEnd_Developer
166
FrontEnd_Developer
167
BackEnd_Developer
BackEnd_Developer
168
169
BackEnd_Developer
BackEnd_Developer
170
171
172
BackEnd_Developer
BackEnd_Developer
173
174
175
BackEnd_D

In [ ]:
data2 = data2[data2.Position.isin(positions)]
data2.head()

Age Gender     City  ... Current Salary Salary one year ago Salary two years ago
0  43      M  München  ...          77000               76200                68000
1  33      F  München  ...          65000               55000                55000
2  32      M  München  ...          88000               73000                54000
3  25      M  München  ...          78000               55000                45000
4  39      M  München  ...          69000               60000                52000

[5 rows x 9 columns]

In [ ]:
data2['Position'].value_counts() 

BackEnd_Developer      426
FrontEnd_Developer      44
Data Scientist          40
FullStack_Developer     24
UX_Designer              8
Name: Position, dtype: int64

In [ ]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 542 entries, 0 to 762
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Age                   542 non-null    object
 1   Gender                542 non-null    object
 2   City                  542 non-null    object
 3   Position              542 non-null    object
 4   Years of experience   542 non-null    object
 5   Your level            542 non-null    object
 6   Current Salary        542 non-null    object
 7   Salary one year ago   542 non-null    object
 8   Salary two years ago  542 non-null    object
dtypes: object(9)
memory usage: 42.3+ KB


In [ ]:
data2.head(10)

Age Gender     City  ... Current Salary Salary one year ago Salary two years ago
0   43      M  München  ...          77000               76200                68000
1   33      F  München  ...          65000               55000                55000
2   32      M  München  ...          88000               73000                54000
3   25      M  München  ...          78000               55000                45000
4   39      M  München  ...          69000               60000                52000
6           M   Berlin  ...          65000               50000                50000
7   36      F  München  ...         109000              105000                97000
8   32      M   Berlin  ...          65000               58000                50000
9   30      M  München  ...          60000               48000                42000
10  40      M   Berlin  ...          70000               57500                54000

[10 rows x 9 columns]

In [ ]:
data2['Years of experience'].value_counts()

10.0    76
5.0     65
8.0     50
6.0     45
7.0     41
4.0     34
12.0    34
3.0     31
9.0     29
11.0    25
15.0    24
2.0     16
1.0     12
13.0    10
20.0     8
14.0     7
18.0     7
2.5      5
         5
16.0     3
17.0     3
1.5      2
30.0     2
0.5      1
7.5      1
19.0     1
0.0      1
25.0     1
4.5      1
38.0     1
21.0     1
Name: Years of experience, dtype: int64

In [ ]:
data2['Current Salary'].unique()

array([77000.0, 65000.0, 88000.0, 78000.0, 69000.0, 109000.0, 60000.0,
       70000.0, 55000.0, 94500.0, 46000.0, 68000.0, 62500.0, 53000.0,
       52000.0, 72000.0, 80000.0, 66000.0, 75800.0, 84000.0, 58000.0,
       113000.0, 75000.0, 57000.0, 50000.0, 71000.0, 56000.0, 30000.0,
       45000.0, 40000.0, 54000.0, 48000.0, 61000.0, 87000.0, 85000.0,
       67000.0, 62000.0, 83000.0, 73000.0, 78600.0, 43000.0, 107000.0,
       64000.0, 95000.0, 42000.0, 74000.0, 72500.0, 102000.0, 55800.0,
       63000.0, 115000.0, 44000.0, 90000.0, 92000.0, 125000.0, 76000.0,
       48500.0, 49000.0, 70500.0, 86000.0, 35000.0, 96000.0, 73300.0,
       73500.0, 130000.0, 47500.0, 81000.0, 53500.0, 87975.0, '',
       140000.0, 31500.0, 120000.0, 112000.0, 100000.0, 59000.0, 47000.0,
       68400.0, 67200.0, 111000.0, 38000.0, 24000.0, 39600.0, 64400.0,
       57800.0, 20400.0, 47300.0, 23162.0, 17532.0, 34000.0, 19200.0,
       51750.0, 38400.0, 51000.0, 69600.0, 58500.0, 47250.0, 64500.0,
       122000

## Remove rows that have empty values in Salary One year ago or Salary two years ago for visulization 2

In [ ]:
data2['Salary two years ago'].unique()

array([68000.0, 55000.0, 54000.0, 45000.0, 52000.0, 50000.0, 97000.0,
       42000.0, 48000.0, '', 34000.0, 86300.0, 44000.0, 63000.0, 40000.0,
       30000.0, 84000.0, 60000.0, 72000.0, 67000.0, 76000.0, 66000.0,
       56000.0, 42500.0, 65000.0, 27000.0, 10001.0, 70000.0, 16000.0,
       57000.0, 43000.0, 58000.0, 12000.0, 69000.0, 75000.0, 32000.0,
       18000.0, 13000.0, 36000.0, 38000.0, 85000.0, 90000.0, 51000.0,
       60600.0, 80000.0, 64000.0, 49000.0, 47000.0, 62000.0, 15000.0,
       61000.0, 20000.0, 120000.0, 26000.0, 22000.0, 53000.0, 100000.0,
       105000.0, 28000.0, 95000.0, 16500.0, 20400.0, 95500.0, 74000.0,
       24000.0, 10800.0, 37050.0, 10500.0, 115000.0, 91000.0, 102000.0,
       27950.0, 37000.0, 98000.0, 44800.0, 39000.0, 41000.0], dtype=object)

In [ ]:
data2.loc[data2['Salary two years ago'] == '']

Age Gender  ... Salary one year ago Salary two years ago
12   31      F  ...                                         
17   24      M  ...               47000                     
26   39      M  ...               74000                     
28   28      M  ...                                         
29   31      M  ...                                         
..   ..    ...  ...                 ...                  ...
724  28      M  ...                                         
743  40      M  ...                                         
745  37      M  ...               53000                     
749  29      M  ...                                         
755  30      M  ...               50000                     

[218 rows x 9 columns]

In [ ]:
data3=data2[data2['Salary two years ago'] != '']

In [ ]:
data3=data2[data2['Salary one year ago'] != '']

In [ ]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 423 entries, 0 to 762
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Age                   423 non-null    object
 1   Gender                423 non-null    object
 2   City                  423 non-null    object
 3   Position              423 non-null    object
 4   Years of experience   423 non-null    object
 5   Your level            423 non-null    object
 6   Current Salary        423 non-null    object
 7   Salary one year ago   423 non-null    object
 8   Salary two years ago  423 non-null    object
dtypes: object(9)
memory usage: 33.0+ KB


In [ ]:
data3.loc[(data3['Salary one year ago'] == '') ]

Empty DataFrame
Columns: [Age, Gender, City, Position, Years of experience, Your level, Current Salary, Salary one year ago, Salary two years ago]
Index: []

In [ ]:
data3.loc[(data3['Salary two years ago'] == '')]

Age Gender     City  ... Current Salary Salary one year ago Salary two years ago
17   24      M  Hamburg  ...          60000               47000                     
26   39      M  München  ...          77000               74000                     
30   39      M   Berlin  ...          75800               66000                     
36   31      M  München  ...          70000               60000                     
44   24      F  München  ...          57000               45000                     
..   ..    ...      ...  ...            ...                 ...                  ...
705  42      M   Berlin  ...          93000               65000                     
707  26      M   Berlin  ...          33000               30000                     
711  31      F   Berlin  ...          42000               42000                     
745  37      M     Köln  ...          53000               53000                     
755  30      M  München  ...          75000               50000                     

[101 rows x 9 columns]

### Save data for first visualization 

In [ ]:
data2.to_csv('./visOne.csv')

In [ ]:
data3.to_csv('./visTwo.csv')